aten/src/ATen/native/cpu/SoftMaxKernel.cpp

```diff
@@ -23,80 +23,88 @@
+  // machine's L1D cache size.
+  int64_t CHUNK_SIZE = std::max(
+      (int64_t)1,
+      (int64_t)(at::internal::GRAIN_SIZE / (sizeof(scalar_t) * dim_size)));
 
-  parallel_for(
-      0,
-      outer_size,
-      grain_size,
-      [&](int64_t begin, int64_t end) {
-        // NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
-        scalar_t tmp_sum_scalar[CHUNK_SIZE];
-        // NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
-        scalar_t max_input_arr[CHUNK_SIZE];
-        for (int64_t ii = begin; ii < end; ii += CHUNK_SIZE) {
-          int64_t loop_end = CHUNK_SIZE;
-          if (ii + CHUNK_SIZE > end)
-            loop_end = end - ii;
-          for (const auto j : c10::irange(loop_end)) {
-            int64_t i = ii + j;
-            scalar_t* input_data = input_data_base + i * dim_size;
-            max_input_arr[j] = vec::reduce_all<scalar_t>(
-                [](Vec& x, Vec& y) { return vec::maximum(x, y); },
-                input_data,
-                dim_size);
-          }
-          for (const auto j : c10::irange(loop_end)) {
-            int64_t i = ii + j;
-            scalar_t* input_data = input_data_base + i * dim_size;
-            scalar_t max_input = max_input_arr[j];
-            tmp_sum_scalar[j] = vec::map_reduce_all<scalar_t>(
-                [max_input](Vec x) { return (x - Vec(max_input)).exp(); },
-                [](Vec x, Vec y) { return x + y; },
-                input_data,
-                dim_size);
-          }
-          // See [Note AVX-SSE transitions] for why this should call the
-          // vectorized version (aside from perf improvements).
-          vec::map(
-              [](Vec x) { return x.log(); },
-              tmp_sum_scalar,
-              tmp_sum_scalar,
-              loop_end);
-          for (const auto j : c10::irange(loop_end)) {
-            int64_t i = ii + j;
-            scalar_t* input_data = input_data_base + i * dim_size;
-            scalar_t* output_data = output_data_base + i * dim_size;
-            scalar_t tmp_sum = tmp_sum_scalar[j];
-            scalar_t max_input = max_input_arr[j];
-
-            // It's necessary to keep the order of the operations below.
-            // In some cases that input is large digits and the difference
-            // is small, if we compute `max_input` plus `tmp_sum` before,
-            // there would be a numerical problem. See an example in
-            // https://github.com/pytorch/pytorch/issues/11752#issuecomment-422883379
-            vec::map(
-                [tmp_sum, max_input](Vec x) { return x - Vec(max_input) - Vec(tmp_sum); },
-                output_data,
-                input_data,
-                dim_size);
-          }
-        }
-      });
+  // we assume that at::GRAIN_SIZE is an appropriate grain-size.
+  int64_t grain_size = std::max(
+      (Vec::size() * internal::GRAIN_SIZE) / (16 * dim_size),
+      (int64_t)1);
```

Comment:
A `grain_size` of 0 is explicitly supported by `parallel_for` to avoid needing to do this `max`. It's behaves the same as `grain_size=1`.

Reply:
Oh, I see, thanks!

I just noticed that I had misread `TORCH_INTERNAL_ASSERT_DEBUG_ONLY(grain_size >= 0)` as `TORCH_INTERNAL_ASSERT_DEBUG_ONLY(grain_size > 0)`.